internal/credential/vault/private_library.go

```diff
@@ -5,91 +5,146 @@ import (
 	"database/sql"
 	"fmt"
 	"strings"
+	"time"
 
 	"github.com/hashicorp/boundary/internal/credential"
 	"github.com/hashicorp/boundary/internal/db/timestamp"
 	"github.com/hashicorp/boundary/internal/errors"
 	"github.com/hashicorp/boundary/internal/kms"
 	wrapping "github.com/hashicorp/go-kms-wrapping"
 	"github.com/hashicorp/go-kms-wrapping/structwrapping"
+	vault "github.com/hashicorp/vault/api"
 	"google.golang.org/protobuf/proto"
 )
 
-var _ credential.Dynamic = (*actualCredential)(nil)
+var _ credential.UserPassword = (*usrPassCred)(nil)
+
+type usrPassCred struct {
+	*baseCred
+	username string
+	password credential.Password
+}
+
+func (c *usrPassCred) Username() string              { return c.username }
+func (c *usrPassCred) Password() credential.Password { return c.password }
+
+var _ credential.Dynamic = (*baseCred)(nil)
+
+type baseCred struct {
+	*Credential
 
-type actualCredential struct {
-	id         string
-	sessionId  string
 	lib        *privateLibrary
 	secretData map[string]interface{}
-	purpose    credential.Purpose
 }
 
-func (ac *actualCredential) GetPublicId() string           { return ac.id }
-func (ac *actualCredential) GetSessionId() string          { return ac.sessionId }
-func (ac *actualCredential) Secret() credential.SecretData { return ac.secretData }
-func (ac *actualCredential) Library() credential.Library   { return ac.lib }
-func (ac *actualCredential) Purpose() credential.Purpose   { return ac.purpose }
+func (bc *baseCred) Secret() credential.SecretData { return bc.secretData }
+func (bc *baseCred) Library() credential.Library   { return bc.lib }
+func (bc *baseCred) Purpose() credential.Purpose   { return bc.lib.Purpose }
+func (bc *baseCred) getExpiration() time.Duration  { return bc.expiration }
+
+// convert converts bc to a specific credential type if bc is not
+// UnspecifiedType.
+func convert(ctx context.Context, bc *baseCred) (dynamicCred, error) {
+	switch bc.Library().CredentialType() {
+	case credential.UserPasswordType:
+		return baseToUsrPass(ctx, bc)
+	}
+	return bc, nil
+}
+
+func baseToUsrPass(ctx context.Context, bc *baseCred) (*usrPassCred, error) {
+	uAttr, pAttr := bc.lib.UsernameAttribute, bc.lib.PasswordAttribute
+	if uAttr == "" {
+		uAttr = "username"
+	}
+	if pAttr == "" {
+		uAttr = "password"
+	}
+
+	var username, password string
+	if u, ok := bc.secretData[uAttr]; ok {
+		username = fmt.Sprintf("%v", u)
+	}
+	if p, ok := bc.secretData[pAttr]; !ok {
```

Comment:
Haha, so I was trying to figure out why the test was passing given the not setting `pAttr` above, the `!ok` should become `ok`.

What this was doing was setting `username` = passwordAttribue and `password` = `<nil>` :)

Reply:
I fixed it and added a bunch of tests.